### PR TITLE
fix(zero): add a flag to limit the uid lease per namespace

### PR DESCRIPTION
**Pull request by NamanJain8**
Fixes DGRAPH-3124

There might be a scenario where in a multi-tenant cluster a malicious user leases out a large number of UIDs or bump it to a large number, causing an issue for the other tenants. 
This PR adds a super flag `--limit` to zero. This flag limits the number of UIDs (`uid-lease=100000;`) that can be leased by a namespace within an interval specified by `refill-interval=30s;`. This PR also changes the xidmap's `New()` API as we need to pass in the namespace and dgo client.

Along with it, it does some minor cleaning where flag parsing was in the critical path.

---
This change is [Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7568)

### CODE REVIEW

**Review by ahsanbarkati — APPROVED** [review]

File: worker/zero_proxy.go [hunk 1]
```diff
@@ -5,10 +5,14 @@ import (
 
 	"github.com/dgraph-io/dgraph/conn"
 	"github.com/dgraph-io/dgraph/protos/pb"
+	"github.com/pkg/errors"
 	"google.golang.org/grpc"
 )
 
 func forwardAssignUidsToZero(ctx context.Context, in *pb.Num) (*pb.AssignedIds, error) {
+	if in.Type != pb.Num_UID {
```
Comment (NamanJain8) [inline worker/zero_proxy.go:13]: Add JWT check here. Just check if the user belongs to some valid namespace. Pass the namespace in context to zero.

Rate limiting in Zero. Probably add a delay by ourselves on zero.
For namespace 0, no limits. Default is infinite.

**Review by manishrjain — CHANGES_REQUESTED** [review]
Looks like there're a lot of changes. So, reduce the scope.

> Reviewed 10 of 18 files at r2, 3 of 3 files at r3, 5 of 6 files at r4.
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7568#-:-MW-i97i50SK2BHpM03g:bf5aqh3)* status: all files reviewed, 14 unresolved discussions (waiting on @NamanJain8 and @vvbalaji-dgraph)

___
*[dgraph/cmd/bulk/loader.go, line 192 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/7568#-MW-dXV5AkaIlk1UlHMr:-MW-dXV5AkaIlk1UlHMs:b-v5qtkx) ([raw file](https://github.com/dgraph-io/dgraph/blob/c9d74493e6ce58a16d4a9055adc88c7af5ff75d7/dgraph/cmd/bulk/loader.go#L192)):*
> ```Go
> 	for {
> 		ctx, cancel := context.WithTimeout(context.Background(), time.Second)
> 		ctx = x.AttachNamespaceOutgoing(ctx, x.GalaxyNamespace)
> ```

We don't need to do this. If Zero doesn't find a namespace, then just assume GalaxyNs.
___
*[dgraph/cmd/live/run.go, line 634 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/7568#-MW-dbY18XdpQDCVvLM-:-MW-dbY18XdpQDCVvLM0:b-v8wx48) ([raw file](https://github.com/dgraph-io/dgraph/blob/c9d74493e6ce58a16d4a9055adc88c7af5ff75d7/dgraph/cmd/live/run.go#L634)):*
> ```Go
> 	x.Checkf(err, "Unable to connect to zero, Is it running at %s?", opt.zero)
> 
> 	creds := z.NewSuperFlag(Live.Conf.GetString("creds")).MergeAndCheckDefault(x.DefaultCreds)
> ```

same over here?
___
*[dgraph/cmd/live/run.go, line 636 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/7568#-MW-dqt7CxgbWS8oLN7N:-MW-dqt7CxgbWS8oLN7O:b3omjuy) ([raw file](https://github.com/dgraph-io/dgraph/blob/c9d74493e6ce58a16d4a9055adc88c7af5ff75d7/dgraph/cmd/live/run.go#L636)):*
> ```Go
> 	creds := z.NewSuperFlag(Live.Conf.GetString("creds")).MergeAndCheckDefault(x.DefaultCreds)
> 	alloc := xidmap.New(xidmap.XidMapOptions{
> 		Zero:      connzero,
> ```

Might want to rename, if this could be a conn to Alpha.
___
*[dgraph/cmd/zero/assign.go, line 184 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/7568#-MW-erB684MoZs7f31E-:-MW-erB684MoZs7f31E0:b-b7yoq2) ([raw file](https://github.com/dgraph-io/dgraph/blob/c9d74493e6ce58a16d4a9055adc88c7af5ff75d7/dgraph/cmd/zero/assign.go#L184)):*
> ```Go
> 	defer span.End()
> 
> 	validateAndGetToken := func() error {
> ```

rateLimit
___
*[dgraph/cmd/zero/assign.go, line 186 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/7568#-MW-dzcF1UoLHnIHNDG1:-MW-dzcF1UoLHnIHNDG2:b1hk8zp) ([raw file](https://github.com/dgraph-io/dgraph/blob/c9d74493e6ce58a16d4a9055adc88c7af5ff75d7/dgraph/cmd/zero/assign.go#L186)):*
> ```Go
> 	validateAndGetToken := func() error {
> 		if num.GetType() != pb.Num_UID {
> 			// We only rate limit lease of UIDs.
> ```

same thing would apply to timestamp too.
___
*[dgraph/cmd/zero/assign.go, line 194 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/7568#-MW-e8gS53jUwDMPs9ka:-MW-e8gS53jUwDMPs9kb:b4p2pwg) ([raw file](https://github.com/dgraph-io/dgraph/blob/c9d74493e6ce58a16d4a9055adc88c7af5ff75d7/dgraph/cmd/zero/assign.go#L194)):*
> ```Go
> 			return err
> 		}
> 		if num.Val > opts.limit.GetUint64("uid-lease") {
> ```

Parse it once, and use it. Don't parse  it again and again.
___
*[dgraph/cmd/zero/assign.go, line 201 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/7568#-MW-eRWh5D7OiS3_xKUf:-MW-eRWh5D7OiS3_xKUg:b-2je60y) ([raw file](https://github.com/dgraph-io/dgraph/blob/c9d74493e6ce58a16d4a9055adc88c7af5ff75d7/dgraph/cmd/zero/assign.go#L201)):*
> ```Go
> 		if !s.rateLimiter.Allow(ns, num.Val) {
> 			// Return error after random delay.
> 			delay := rand.Intn(int(opts.limit.GetDuration("refill-interval")))
> ```

parse once. For rand, use a locally generated rand. rand.NewSource rand.NewRand or something.
___
*[dgraph/cmd/zero/assign.go, line 203 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/7568#-MW-eif3Ew-MpFWxyZk_:-MW-eif3Ew-MpFWxyZka:b-csudww) ([raw file](https://github.com/dgraph-io/dgraph/blob/c9d74493e6ce58a16d4a9055adc88c7af5ff75d7/dgraph/cmd/zero/assign.go#L203)):*
> ```Go
> 			delay := rand.Intn(int(opts.limit.GetDuration("refill-interval")))
> 			time.Sleep(time.Duration(delay) * time.Second)
> 			return errors.New("Cannot lease UID because UID lease for the namespace is exhausted." +
> ```

for namespace: %#x is exhausted.
___
*[dgraph/cmd/zero/assign.go, line 236 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/7568#-MW-f1D-35lyj4ZLcPcC:-MW-f1D-35lyj4ZLcPcD:b-x6bf54) ([raw file](https://github.com/dgraph-io/dgraph/blob/c9d74493e6ce58a16d4a9055adc88c7af5ff75d7/dgraph/cmd/zero/assign.go#L236)):*
> ```Go
> 		// pass on the incoming metadata to the zero leader.
> 		if md, ok := metadata.FromIncomingContext(ctx); ok {
> 			ctx = metadata.NewOutgoingContext(ctx, md)
> ```

Do you need to do this?
___
*[dgraph/cmd/zero/http.go, line 71 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/7568#-MW-g4MkAWHgBzSBt0Yy:-MW-g4MkAWHgBzSBt0Yz:b-dc3t57) ([raw file](https://github.com/dgraph-io/dgraph/blob/c9d74493e6ce58a16d4a9055adc88c7af5ff75d7/dgraph/cmd/zero/http.go#L71)):*
> ```Go
> 	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
> 	// Consider all the request coming through /assign belong to GalaxyNamespace.
> 	ctx = x.AttachNamespace(ctx, x.GalaxyNamespace)
> ```

can remove.
___
*[dgraph/cmd/zero/run.go, line 109 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/7568#-MW-gCqm4SQNx_D1Q3C6:-MW-gCqm4SQNx_D1Q3C7:bwam1i7) ([raw file](https://github.com/dgraph-io/dgraph/blob/c9d74493e6ce58a16d4a9055adc88c7af5ff75d7/dgraph/cmd/zero/run.go#L109)):*
> ```Go
> 			in an interval specified by refill-interval.`).
> 		Flag("refill-interval",
> 			"The interval after which the tokens for UID lease are replenished,").
> ```

comma at the end.
___
*[x/x.go, line 1413 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/7568#-MW-gXbh6ZYLUWt08AVQ:-MW-gXbh6ZYLUWt08AVR:bezjgrj) ([raw file](https://github.com/dgraph-io/dgraph/blob/c9d74493e6ce58a16d4a9055adc88c7af5ff75d7/x/x.go#L1413)):*
> ```Go
> 
> // RateLimiter implements a basic rate limiter.
> type RateLimiter struct {
> ```

There are good libraries which exist already.
___
*[x/x.go, line 1425 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/7568#-MW-h5Q93NwQCabb-qtm:-MW-h5Q93NwQCabb-qtn:byanx07) ([raw file](https://github.com/dgraph-io/dgraph/blob/c9d74493e6ce58a16d4a9055adc88c7af5ff75d7/x/x.go#L1425)):*
> ```Go
> 	r := &RateLimiter{
> 		limiter: lockedMap{
> 			mp: make(map[uint64]uint64),
> ```

Use sync.Map for the map, you store the key and the pointer. And use atomics for the values.

And then also try maybe current approach. Do some parallel benchmarking, see which one is faster. In your benchmarks assume that all the namespaces that need to be accessed are already in the map.

Also benchmark with 3rd party solutions.

**Review by NamanJain8 — COMMENTED** [review]
> Dismissed @manishrjain from 7 discussions.
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7568)* status: 2 of 17 files reviewed, 7 unresolved discussions (waiting on @manishrjain, @NamanJain8, @pawanrawal, and @vvbalaji-dgraph)

___
*[dgraph/cmd/bulk/loader.go, line 192 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/7568#-MW-dXV5AkaIlk1UlHMr:-MWZzWHwEToDmeeoZCPY:b-i21bbo) ([raw file](https://github.com/dgraph-io/dgraph/blob/c9d74493e6ce58a16d4a9055adc88c7af5ff75d7/dgraph/cmd/bulk/loader.go#L192)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

We don't need to do this. If Zero doesn't find a namespace, then just assume GalaxyNs.
</blockquote></details>

Thanks. Done.
___
*[dgraph/cmd/live/run.go, line 634 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/7568#-MW-dbY18XdpQDCVvLM-:-MW_L0v48h_nrGuqjPW8:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/c9d74493e6ce58a16d4a9055adc88c7af5ff75d7/dgraph/cmd/live/run.go#L634)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

same over here?
</blockquote></details>

Done.
___
*[dgraph/cmd/live/run.go, line 636 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/7568#-MW-dqt7CxgbWS8oLN7N:-MW_LzRS3W1Ru6Jgi0h9:b-2e35kr) ([raw file](https://github.com/dgraph-io/dgraph/blob/c9d74493e6ce58a16d4a9055adc88c7af5ff75d7/dgraph/cmd/live/run.go#L636)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Might want to rename, if this could be a conn to Alpha.
</blockquote></details>

Renamed it to `UidAssigner`
___
*[dgraph/cmd/zero/assign.go, line 184 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/7568#-MW-erB684MoZs7f31E-:-MW_-0ab8nGP_XGXYdl3:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/c9d74493e6ce58a16d4a9055adc88c7af5ff75d7/dgraph/cmd/zero/assign.go#L184)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

rateLimit
</blockquote></details>

Done.
___
*[dgraph/cmd/zero/assign.go, line 194 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/7568#-MW-e8gS53jUwDMPs9ka:-MW_-2Yp5zH3nXGtepG5:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/c9d74493e6ce58a16d4a9055adc88c7af5ff75d7/dgraph/cmd/zero/assign.go#L194)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Parse it once, and use it. Don't parse  it again and again.
</blockquote></details>

Done.
___
*[dgraph/cmd/zero/assign.go, line 201 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/7568#-MW-eRWh5D7OiS3_xKUf:-MW_aSNH4nEVRxHLRCTy:b-6a603j) ([raw file](https://github.com/dgraph-io/dgraph/blob/c9d74493e6ce58a16d4a9055adc88c7af5ff75d7/dgraph/cmd/zero/assign.go#L201)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

parse once. For rand, use a locally generated rand. rand.NewSource rand.NewRand or something.
</blockquote></details>

Not sure about the benefit of it. As the local source is thread-unsafe and we will need to have some extra mechanism to do locking and stuff. 
___
*[dgraph/cmd/zero/assign.go, line 203 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/7568#-MW-eif3Ew-MpFWxyZk_:-MWZzlohEQy5u1hTgYc5:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/c9d74493e6ce58a16d4a9055adc88c7af5ff75d7/dgraph/cmd/zero/assign.go#L203)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

for namespace: %#x is exhausted.
</blockquote></details>

Done.
___
*[dgraph/cmd/zero/assign.go, line 236 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/7568#-MW-f1D-35lyj4ZLcPcC:-MW_JihO1B1dGPfVeSL1:bdf9xax) ([raw file](https://github.com/dgraph-io/dgraph/blob/c9d74493e6ce58a16d4a9055adc88c7af5ff75d7/dgraph/cmd/zero/assign.go#L236)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Do you need to do this?
</blockquote></details>

Yes, this is needed because the metadata from the incoming context is lost when the context is passed to some other server. 
___
*[dgraph/cmd/zero/http.go, line 71 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/7568#-MW-g4MkAWHgBzSBt0Yy:-MW_Jteq6mNB5OjigFMc:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/c9d74493e6ce58a16d4a9055adc88c7af5ff75d7/dgraph/cmd/zero/http.go#L71)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

can remove.
</blockquote></details>

Done.
___
*[dgraph/cmd/zero/run.go, line 109 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/7568#-MW-gCqm4SQNx_D1Q3C6:-MW_JwdC0EGB_LavW7av:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/c9d74493e6ce58a16d4a9055adc88c7af5ff75d7/dgraph/cmd/zero/run.go#L109)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

comma at the end.
</blockquote></details>

Done.
___
*[x/x.go, line 1425 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/7568#-MW-h5Q93NwQCabb-qtm:-MW_KWPEBt8oVrUM7Ms2:b-qd3bs5) ([raw file](https://github.com/dgraph-io/dgraph/blob/c9d74493e6ce58a16d4a9055adc88c7af5ff75d7/x/x.go#L1425)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Use sync.Map for the map, you store the key and the pointer. And use atomics for the values.

And then also try maybe current approach. Do some parallel benchmarking, see which one is faster. In your benchmarks assume that all the namespaces that need to be accessed are already in the map.

Also benchmark with 3rd party solutions.
</blockquote></details>

Done benchmarking various approaches (See https://gist.github.com/NamanJain8/5a935fb33cc43b6175d7da66cc7ead70). Sync map with atomics works best. Couldn't find any third-party rate limiter that fits our use case.

**Review by manishrjain — APPROVED** [review]
:lgtm:

> Reviewed 17 of 21 files at r5.
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7568#-:-MWeYTb6E6u3u4vMz02q:bnfp4nl)* status: all files reviewed, 10 unresolved discussions (waiting on @manishrjain, @NamanJain8, @pawanrawal, and @vvbalaji-dgraph)

___
*[dgraph/cmd/alpha/run.go, line 735 at r5](https://reviewable.io/reviews/dgraph-io/dgraph/7568#-MWeXCqX79ZoDNixC06J:-MWeXCqX79ZoDNixC06K:b-mazge3) ([raw file](https://github.com/dgraph-io/dgraph/blob/02bb49fae5af45fc9b1459b049b531acce9a2950/dgraph/cmd/alpha/run.go#L735)):*
> ```Go
> 	x.Config.LimitQueryEdge = x.Config.Limit.GetUint64("query-edge")
> 	x.Config.BlockClusterWideDrop = x.Config.Limit.GetBool("disallow-drop")
> 	x.Config.LimitNormalizeNode = int(x.Config.Limit.GetInt64("normalize-node"))
> ```

try to keep this local.
___
*[x/x.go, line 470 at r5](https://reviewable.io/reviews/dgraph-io/dgraph/7568#-MWeY4vB3R2FpQc50XnE:-MWeY4vB3R2FpQc50XnF:b-mhtu95) ([raw file](https://github.com/dgraph-io/dgraph/blob/02bb49fae5af45fc9b1459b049b531acce9a2950/x/x.go#L470)):*
> ```Go
> // AttachJWTNamespaceOutgoing attaches the namespace in the JWT claims to the outgoing metadata of
> // the context.
> func AttachJWTNamespaceOutgoing(ctx context.Context) (context.Context, error) {
> ```

`ContextFromJWT`
___
*[x/x.go, line 482 at r5](https://reviewable.io/reviews/dgraph-io/dgraph/7568#-MWeY2-t-Geav8XyEDqB:-MWeY2-t-Geav8XyEDqC:bako04g) ([raw file](https://github.com/dgraph-io/dgraph/blob/02bb49fae5af45fc9b1459b049b531acce9a2950/x/x.go#L482)):*
> ```Go
> 
> // AttachNamespaceOutgoing adds given namespace in the outgoing metadata of the context.
> func AttachNamespaceOutgoing(ctx context.Context, namespace uint64) context.Context {
> ```

`func ContextFromNamespace(...)`

File: x/x.go [hunk 2]
```diff
@@ -1380,8 +1406,81 @@ func PrefixesToMatches(prefixes [][]byte, ignore string) []*pb.Match {
...
+func (r *RateLimiter) RefillPeriodically() {
+	defer r.closer.Done()
+	refill := func() {
+		r.limiter.Range(func(_, val interface{}) bool {
+			atomic.StoreInt64(val.(*int64), r.maxTokens)
+			return true
+		})
+	}
+
+	ticker := time.NewTicker(r.refillAfter)
```
Comment (jarifibrahim) [inline x/x.go:1460]: defer ticker.Stop()

**Review by jarifibrahim — APPROVED** [review]

File: x/x.go (hunk 2 above)
Comment (NamanJain8) [reply x/x.go:1460]: Thanks done.

**Pull request merged**